parser/parser.coffee

```diff
@@ -0,0 +1,235 @@
+scanner = require '../scanner/scanner'
+CustomError = require '../error/custom_error'
+
+Program = require './entities/program'
+Block = require './entities/block'
+ForStatement = require './entities/for_statement'
+WhileStatement = require './entities/while_statement'
+IntegerLiteral = require './entities/integer_literal'
+FloatLiteral = require './entities/float_literal'
+StringLiteral = require './entities/string_literal'
+BooleanLiteral = require './entities/boolean_literal'
+VariableReference = require './entities/variable_reference'
+VariableDeclaration = require './entities/variable_declaration'
+VariableAssignment = require './entities/variable_assignment'
+ListLiteral = require './entities/list_literal'
+SetLiteral = require './entities/set_literal'
+UnaryExpression = require './entities/unary_expression'
+BinaryExpression = require './entities/binary_expression'
+Tokens = require '../scanner/tokens'
+StartTokens = require './start_tokens'
+
+tokens = []
+
+module.exports = (scannerOutput) ->
+  tokens = scannerOutput
+  program = parseProgram()
+  match 'EOF'
+  program
+
+parseProgram = ->
+  new Program parseBlock()
+
+parseBlock = ->
+  statements = []
+  loop
+    statements.push parseStatement() unless at 'newline'
+    match 'newline' while at 'newline'
+    break unless at StartTokens.expression
+  new Block statements
+
+parseStatement = ->
+  if at 'for'
+    parseForLoop()
+  else if at 'while'
+    parseWhileLoop()
+  else
+    parseExpression()
+
+parseForLoop = ->
+  match 'for'
+  id = match 'ID'
+  match 'in'
+  iterable = parseExpression()
+  match ':'
+  if at 'newline'
+    body = parseBlock()
+    match 'end'
+  else
+    body = parseExpression()
+    match 'newline'
+  new ForStatement id, iterable, body
+
+parseWhileLoop = ->
+  match 'while'
+  condition = parseExpression()
+  match ':'
+  if at 'newline'
+    body = parseBlock()
+    match 'end'
+  else
+    body = parseExpression()
+    match 'newline'
+  new WhileStatement condition, body
+
+parseVarDec = ->
+  id = match 'ID'
+  match ':='
+  # TODO: match type?
+  exp = parseExpression()
+  new VariableDeclaration id, exp
+
+parseVarAssig = ->
+  id = match 'ID'
+  match '='
+  exp = parseExpression()
+  new VariableAssignment id, exp
+
+parseConditional = ->
+  if at 'if'
```

Comment:
I just used your parser code for `if` statements. I didn't have the chance to add anything to it yet,